src/Microsoft.AspNet.Buffering/BufferingWriteStream.cs

```diff
@@ -0,0 +1,186 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.Buffering
+{
+    internal class BufferingWriteStream : Stream
+    {
+        private readonly Stream _innerStream;
+        private readonly MemoryStream _buffer = new MemoryStream();
```

Comment:
Might want to create a new issue to adjust this a bit eventually to accommodate arbitrarily large responses. For example, this could be a HybridStream that keeps the first 64k in memory (like in non-linear 8k memory pooled segments), and buffers any data over that size in a temp file (prefereable delete-on-close).